bot/utils/persistent_scheduling.py

```diff
@@ -0,0 +1,161 @@
+import asyncio
+import typing as t
+from datetime import datetime, timedelta, timezone
+
+import arrow
+from arrow import Arrow
+from async_rediscache import RedisCache
+
+from bot.bot import Bot
+from bot.utils.scheduling import Scheduler
+
+SchedulerTaskFactory = t.Callable[[t.Union[str, int]], t.Coroutine[None, None, None]]
+CacheKey = t.Union[str, int]
+
+
+class PersistentScheduler:
+    """
+    Schedule the execution of coroutines and keep track of them even across restarts and cog reloads.
+
+    When instantiating a PersistentScheduler, a name must be provided. This name is used to distinguish the
+    instance's log messages and RedisCache from other instances. Using the name of the class or module containing
+    the instance is suggested. The same name must be provided after a restart for the tasks to be preserved.
+
+    A coroutine must be provided in order to reschedule each of the keys contained in the cache at init.
+    The coroutine will take a single argument -- the ID that was used to create the Scheduler task in the first place.
+
+    Coroutines can be scheduled with `schedule_at` or `schedule_later`.
+    A unique ID is required to be given in order to keep track of the resulting Tasks.
+    Any scheduled task can be cancelled prematurely using `cancel` by providing the same ID used to schedule it, but
+    the task will be rescheduled from the cache on the next initialization. To remove the task completely, you can
+    `delete` it.
+    The `in` operator is supported for checking if a task with a given ID is currently scheduled.
+
+    Any exception raised in a scheduled task is logged when the task is done.
+    """
+
+    def __init__(self, name: str, task_factory: SchedulerTaskFactory, bot: Bot):
+        name = f"{__name__}.{name}"
+
+        self._scheduler = Scheduler(name)
```

Comment:
Ah, yes. The reason I did this is that I wanted both the scheduler and the cache to have the same name. Hopefully it's not too much of an issue. A way to around it is to add an optional namespace argument to the scheduler init, although I'm not sure if that's desirable.

Reply:
> Hopefully it's not too much of an issue
The logger name will be really long since `__name__` is full qualified and two fully qualified names are being concatenated.

> A way to around it is to add an optional namespace argument to the scheduler init, although I'm not sure if that's desirable.

I'd be okay with it if it was really necessary, but I don't think it's important to keep the names the same. In fact, it may be a bad idea to use `__name__` for the cache, since it will cause cache misses if there's a refactor that moves modules around or renames them.

Reply:
It would be nice to be able to refer to both using the same name for debugging purposes, although it's not a must.